collated/test/A0147971U.md

```diff
@@ -0,0 +1,221 @@
+        commandBox.runCommand("undo");
+        commandBox.runCommand("delete 1");
+        assertRedoResult("redo", RedoCommand.MESSAGE_NO_PAST_COMMAND_TO_REDO,
+                td.submitProgressReport, td.developerMeeting,
+                td.researchWhales, td.learnVim, 
+                td.buyBirthdayGift, td.signUpForYoga);
+    }
+    
+    //---------------- redo (positive integer) ----------------------
+    
+    @Test
+    public void redo_four_hasThreeHistory_redoAll() {
+        commandBox.runCommand("delete 1");
+        commandBox.runCommand("delete 1");
+        commandBox.runCommand("delete 1");
+        commandBox.runCommand("undo 3");
+        assertRedoResult("redo 4", createRedoSuccessResultMessage(3), 
+                td.researchWhales, td.learnVim, 
+                td.buyBirthdayGift, td.signUpForYoga);
+    }
+    
+    @Test
+    public void redo_three_hasThreeHistory_redoThree() {
+        commandBox.runCommand("delete 1");
+        commandBox.runCommand("delete 1");
+        commandBox.runCommand("delete 1");
+        commandBox.runCommand("undo 3");
+        assertRedoResult("redo 3", createRedoSuccessResultMessage(3), 
+                td.researchWhales, td.learnVim, 
+                td.buyBirthdayGift, td.signUpForYoga);
+    }
+    
+    @Test
+    public void redo_three_hasThreeHistory_redoOneByOneUntilOriginal() {
+        commandBox.runCommand("delete 1");
+        commandBox.runCommand("delete 1");
+        commandBox.runCommand("delete 1");
+        commandBox.runCommand("undo 3");
+        assertRedoResult("redo 1", createRedoSuccessResultMessage(1), 
+                td.submitProgressReport, td.developerMeeting,
+                td.researchWhales, td.learnVim, 
+                td.buyBirthdayGift, td.signUpForYoga);
+        assertRedoResult("redo 1", createRedoSuccessResultMessage(1), 
+                td.developerMeeting,
+                td.researchWhales, td.learnVim, 
+                td.buyBirthdayGift, td.signUpForYoga);
+        assertRedoResult("redo 1", createRedoSuccessResultMessage(1), 
+                td.researchWhales, td.learnVim, 
+                td.buyBirthdayGift, td.signUpForYoga);
+        assertRedoResult("redo 1", RedoCommand.MESSAGE_NO_PAST_COMMAND_TO_REDO,
+                td.researchWhales, td.learnVim, 
+                td.buyBirthdayGift, td.signUpForYoga);
+    }
+    
+    @Test
+    public void redo_three_hasZeroHistory_remainsUnchanged() {
+        assertRedoResult("redo 3", RedoCommand.MESSAGE_NO_PAST_COMMAND_TO_REDO,
+                td.submitPrototype, 
+                td.submitProgressReport, td.developerMeeting,
+                td.researchWhales, td.learnVim, 
+                td.buyBirthdayGift, td.signUpForYoga);
+    }
+    
+    //---------------- redo (illegal arguments) ----------------------
+    
+    @Test
+    public void redo_nonpositiveInteger_hasThreeHistory_remainsUnchanged_errorMessageShown() {
+        commandBox.runCommand("delete 1");
```

Comment:
you can consider moving all this to a `setup()` method and eliminate code duplication.